File: posthog/helpers/semver.py
import re
from typing import Optional, Union, Any
from collections.abc import Callable
from dataclasses import dataclass
from functools import cmp_to_key


@dataclass
class SemanticVersion:
    major: int
    minor: Optional[int] = None
    patch: Optional[int] = None
    extra: Optional[str] = None


@dataclass
class VersionDiff:
    kind: str
    diff: int


def parse_version(version: str) -> SemanticVersion:
    """Parse a version string into a SemanticVersion object.

    Raises ValueError if the version string is invalid.
    """
    # Split on hyphen to extract extra
    split = version.split("-", 1)
    version_part = split[0]
    extra = split[1] if len(split) > 1 else None

    # Match the version pattern
    match = re.match(r"^v?(\d+)(?:\.(\d+))?(?:\.(\d+))?$", version_part)
    if not match:
        raise ValueError(f"Invalid semver string: {version}")

    major_str, minor_str, patch_str = match.groups()

    major = int(major_str)
    minor = int(minor_str) if minor_str is not None else None
    patch = int(patch_str) if patch_str is not None else None

    return SemanticVersion(major=major, minor=minor, patch=patch, extra=extra)


def try_parse_version(version: str) -> Optional[SemanticVersion]:
    """Try to parse a version string, return None if it's invalid."""
    try:
        return parse_version(version)
    except ValueError:
        return None


def diff_versions(a: Union[str, SemanticVersion], b: Union[str, SemanticVersion]) -> Optional[VersionDiff]:
    """Compare two versions and return the difference, or None if they're equal."""
    try:
        pa = parse_version(a) if isinstance(a, str) else a
        pb = parse_version(b) if isinstance(b, str) else b

        if pa.major != pb.major:
            return VersionDiff(kind="major", diff=pa.major - pb.major)

        a_minor = pa.minor or 0
        b_minor = pb.minor or 0
        if a_minor != b_minor:
            return VersionDiff(kind="minor", diff=a_minor - b_minor)

        a_patch = pa.patch or 0
        b_patch = pb.patch or 0
        if a_patch != b_patch:
            return VersionDiff(kind="patch", diff=a_patch - b_patch)

        if pa.extra != pb.extra:
            # not having an extra is treated as a higher version than having an extra
            if pa.extra:
                if pb.extra:
                    # Simple string comparison
                    return VersionDiff(kind="extra", diff=1 if pa.extra > pb.extra else -1)
                return VersionDiff(kind="extra", diff=-1)
            if pb.extra:
                return VersionDiff(kind="extra", diff=1)
            return None

        return None
    except ValueError:
        return None


def compare_version(a: Union[str, SemanticVersion], b: Union[str, SemanticVersion]) -> int:
    """Compare two versions and return the difference as an integer.

    Returns:
        - Positive if a > b
        - Zero if a == b
        - Negative if a < b
    """
    diff = diff_versions(a, b)
    if not diff:
        return 0
    return diff.diff


def lowest_version(versions: list[Union[str, SemanticVersion]]) -> SemanticVersion:
    """Find the lowest version in a list."""
    parsed = [parse_version(v) if isinstance(v, str) else v for v in versions]
    # Use compare_version which properly handles extra components
    return min(parsed, key=cmp_to_key(compare_version))


def highest_version(versions: list[Union[str, SemanticVersion]]) -> SemanticVersion:
    """Find the highest version in a list."""
    parsed = [parse_version(v) if isinstance(v, str) else v for v in versions]
    # Use compare_version which properly handles extra components
    return max(parsed, key=cmp_to_key(compare_version))


def is_equal_version(a: Union[str, SemanticVersion], b: Union[str, SemanticVersion]) -> bool:
    """Check if two versions are equal."""
    return diff_versions(a, b) is None


def version_to_string(version: Union[str, SemanticVersion, Any]) -> str:
    """Convert a version object to a string."""
    if isinstance(version, str):
        return version.lstrip("v")

    if isinstance(version, SemanticVersion):
        version_part = f"{version.major}"
        if version.minor is not None:
            version_part += f".{version.minor}"
            if version.patch is not None:
                version_part += f".{version.patch}"

        if version.extra:
            return f"{version_part}-{version.extra}"

        return version_part

    # Fall back to string conversion for other types
    return str(version).lstrip("v")


def create_version_checker(
    required_version: Union[str, SemanticVersion],
) -> Callable[[Union[str, SemanticVersion]], bool]:
    """Create a function that checks if a version is at least the required version."""

    def check_version(version: Union[str, SemanticVersion]) -> bool:
        diff = diff_versions(version, required_version)
        return diff is None or diff.diff > 0

    return check_version

File: .conflict-side-1/frontend/src/lib/utils/semver.ts
export interface SemanticVersion {
    major: number
    minor?: number
    patch?: number
    extra?: string
}

export function parseVersion(version: string): SemanticVersion {
    // Regex to parse a semver string, can be just major, major.minor or major.minor.patch
    // if there is a hyphen thing ignore anything afterwards
    // strip a leading v if it exists
    const split = version.split('-', 2)
    version = split[0]
    const extra = split[1]
    const match = version.match(/^v?(\d+)(?:\.(\d+))?(?:\.(\d+))?$/)
    if (!match) {
        throw new Error(`Invalid semver string: ${version}`)
    }
    const [, major, minor, patch] = match
    const majorInt = parseInt(major, 10)
    const minorInt = minor != null ? parseInt(minor, 10) : undefined
    const patchInt = patch != null ? parseInt(patch, 10) : undefined

    if (isNaN(majorInt) || (minorInt != null && isNaN(minorInt)) || (patchInt != null && isNaN(patchInt))) {
        throw new Error(`Invalid semver string: ${version}`)
    }
    return { major: majorInt, minor: minorInt, patch: patchInt, extra }
}

export function tryParseVersion(version: string): SemanticVersion | null {
    try {
        return parseVersion(version)
    } catch {
        return null
    }
}

export interface SemanticVersionDiff {
    kind: 'major' | 'minor' | 'patch' | 'extra'
    diff: number
}

export function diffVersions(a: string | SemanticVersion, b: string | SemanticVersion): SemanticVersionDiff | null {
    const pa = typeof a === 'string' ? parseVersion(a) : a
    const pb = typeof b === 'string' ? parseVersion(b) : b

    if (pa.major !== pb.major) {
        return { kind: 'major', diff: pa.major - pb.major }
    }
    if (pa.minor !== pb.minor) {
        return { kind: 'minor', diff: (pa.minor ?? 0) - (pb.minor ?? 0) }
    }
    if (pa.patch !== pb.patch) {
        return { kind: 'patch', diff: (pa.patch ?? 0) - (pb.patch ?? 0) }
    }
    if (pa.extra !== pb.extra) {
        // not having an extra is treated as a higher version than having an extra
        if (pa.extra) {
            if (pb.extra) {
                // we don't really use these, if we ever do we should improve this logic to better compare them,
                // for now just do a simple string comparison rather than trying to parse alpha/beta/rc/etc
                return { kind: 'extra', diff: pa.extra.localeCompare(pb.extra) }
            }

            return { kind: 'extra', diff: -1 }
        }
        if (pb.extra) {
            return { kind: 'extra', diff: 1 }
        }
        return null
    }
    return null
}

export function compareVersion(a: string | SemanticVersion, b: string | SemanticVersion): number {
    const diff = diffVersions(a, b)
    if (!diff) {
        return 0
    }
    return diff.diff
}

export function lowestVersion(versions: (string | SemanticVersion)[]): SemanticVersion {
    const parsed = versions.map((v) => (typeof v === 'string' ? parseVersion(v) : v))
    // we expect this list to be small, so don't worry about nlogn vs n from using sort
    parsed.sort(compareVersion)
    return parsed[0]
}

export function highestVersion(versions: (string | SemanticVersion)[]): SemanticVersion {
    const parsed = versions.map((v) => (typeof v === 'string' ? parseVersion(v) : v))
    parsed.sort(compareVersion)
    return parsed[parsed.length - 1]
}

export function isEqualVersion(a: string | SemanticVersion, b: string | SemanticVersion): boolean {
    return diffVersions(a, b) === null
}

export function versionToString(version: SemanticVersion): string {
    const versionPart = `${version.major}${
        version.minor != null ? `.${version.minor}${version.patch != null ? `.${version.patch}` : ''}` : ''
    }`
    if (version.extra) {
        return `${versionPart}-${version.extra}`
    }
    return versionPart
}

export function createVersionChecker(requiredVersion: string | SemanticVersion) {
    return (version: string | SemanticVersion): boolean => {
        const diff = diffVersions(version, requiredVersion)
        return !diff || diff.diff > 0
    }
}
